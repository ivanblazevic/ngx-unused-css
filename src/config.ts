import { FileImporter, Importer } from 'sass';

export interface Ignore {
  file: string;
  all?: boolean;
  selectors?: string[];
}

export type SupportedStyleExtensions = 'scss' | 'sass' | 'css';

export interface Config {
  path: string;
  styleExt?: SupportedStyleExtensions;
  ignore: (string | Ignore)[];
  importer?: Importer<'sync'> | FileImporter<'sync'>;
  includePaths?: string[];
  globalStyles?: string;
  cli: boolean; // Determine if running via CLI or by using programmic API
}
