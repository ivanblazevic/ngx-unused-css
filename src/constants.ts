import { SupportedStyleExtensions } from './config';

export const SELECTORS_TO_IGNORE = [':host', '::ng-deep'];
export const DEFAULT_STYLE_EXTENSION: SupportedStyleExtensions = 'scss';
