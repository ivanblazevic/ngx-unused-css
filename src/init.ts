import chalk from 'chalk';
import fs from 'fs';
import prompts from 'prompts';
import { Config } from './config';

const questions: prompts.PromptObject<string>[] = [
  {
    type: 'confirm',
    name: 'isDefaultSrc',
    message: 'Is your app located in src/app?'
  },
  {
    type: 'confirm',
    name: 'isDefaultGlobalStylingSrc',
    message: 'Is your global styling located in src/styles.scss?'
  },
  {
    type: 'confirm',
    name: 'hasMaterialLib',
    message: 'Does your project use Angular Material library?'
  },
  {
    type: 'select',
    name: 'styleExt',
    message: 'Which style extension is your project using?',
    choices: [
      { title: 'SCSS', value: '.scss' },
      { title: 'SASS', value: '.sass' },
      { title: 'CSS', value: '.css' }
    ],
    initial: 0
  }
];

export default async function init() {
  const result = await prompts(questions);

  const feedback = [];

  const config: Config = {
    path: '',
    ignore: [],
    cli: true
  };

  if (result.isDefaultSrc) {
    config.path = 'src/app';
  } else {
    feedback.push(
      'App is not located in the src/app, pls provide correct path in the config'
    );
  }

  if (result.isDefaultGlobalStylingSrc) {
    config.globalStyles = 'src/styles.scss';
  } else {
    feedback.push(
      'Looks like configuration for global styling is missing, pls provide correct globalStyles in the config if there is any'
    );
  }

  if (result.hasMaterialLib) {
    config.ignore.push('.mat-');
  }

  config.styleExt = result.styleExt;

  if (feedback.length > 0) {
    console.log(
      chalk.red.bold(
        'INFO: Following changes in the .ngx-unused-css.json are required:\n'
      )
    );

    feedback.forEach((f, idx) => {
      console.log(chalk.red.bold(`${idx + 1}. ${f}`));
    });
  }

  fs.writeFileSync('.ngx-unused-css.json', JSON.stringify(config, null, 2));

  process.exit(0);
}
