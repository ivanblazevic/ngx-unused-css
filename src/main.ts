import chalk from 'chalk';
import { table } from 'table';
import { Config } from './config';
import { DEFAULT_STYLE_EXTENSION } from './constants';
import { UnusedClassesMap } from './helpers/unusedClassMapper';
import UnusedClasses from './main/getUnusedClasses';

class Main {
  constructor(config: Config) {
    if (!config.styleExt) {
      config.styleExt = DEFAULT_STYLE_EXTENSION;
    }

    const unusedClasses = new UnusedClasses(config);

    unusedClasses.getUnusedClasses(config.path).then((res) => {
      if (config.globalStyles) {
        unusedClasses.getGlobalUnusedClasses(config.globalStyles).then((r) => {
          if (r.length > 0) {
            // @ts-ignore
            res.push([r, '***** GLOBAL UNUSED CSS *****']);
          }
          if (res.length > 0) {
            this.log(res);
          }
        });
      } else {
        if (res.length > 0) {
          this.log(res);
        }
      }
    });
  }

  private log(classes: UnusedClassesMap[]) {
    let result = chalk.red.bold(
      'Unused CSS classes were found for the following files:\n\n'
    );

    classes.forEach((e: [string[], string]) => {
      const htmlPath = e[1];
      const cssPath = e[1].replace('.html', '.scss');

      result += chalk.red(htmlPath) + '\n';
      result += chalk.red.bold(cssPath) + '\n';

      const cssClasses = e[0].join('\n');

      result += table([[chalk.green(cssClasses)]]);
    });

    console.log(result);

    process.exit(1);
  }
}

export default Main;
