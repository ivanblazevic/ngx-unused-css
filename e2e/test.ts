import { exec, execSync } from 'child_process';
import { Config } from '../src/config';
import Main from './../src/main';

describe('e2e', () => {
  it('should test CLI', (done) => {
    exec(
      'node ./../dist/index.js',
      { cwd: 'e2e' },
      function (error, stdout, stderr) {
        if (error) {
          expect(stdout).toMatchSnapshot();
          done();
        }
      }
    );
  });

  it('should test API', async () => {
    const config: Config = require('./.ngx-unused-css.json');

    // Copy test libs to root node modules to make tests not failing
    execSync('cp -R e2e/node_modules/test-css-lib node_modules/test-css-lib');

    // Override paths
    config.path = 'e2e/src/app';
    config.globalStyles = 'e2e/src/styles.scss';
    config.includePaths = ['e2e/node_modules/test-css-lib-load-paths'];

    const instance = new Main(config);

    const result = await instance.run();

    expect(result).toEqual({
      css: [
        [
          ['.test-lib-class', '.test-css-lib-load-paths', '.unused'],
          'e2e/src/app/app.component.html'
        ]
      ],
      globalCss: ['.i-am-global-class']
    });
  });
});
