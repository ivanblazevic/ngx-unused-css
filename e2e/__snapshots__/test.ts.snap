// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`e2e should test CLI 1`] = `
"Unused CSS classes were found for the following files:

src/app/app.component.html
src/app/app.component.scss
╔══════════════════════════╗
║ .test-lib-class          ║
║ .test-css-lib-load-paths ║
║ .unused                  ║
╚══════════════════════════╝
***** GLOBAL UNUSED CSS *****
***** GLOBAL UNUSED CSS *****
╔════════════════════╗
║ .i-am-global-class ║
╚════════════════════╝

"
`;
